action/protocol/staking/protocol.go

```diff
@@ -114,3 +147,55 @@ func (p *Protocol) handleRestake(ctx context.Context, act action.Action, sm prot
 	// TODO
 	return nil, nil
 }
+
+func (p *Protocol) hasDelegate(name CandName) bool {
+	return p.delegates.Contains(name)
+}
+
+func (p *Protocol) validateDelegate(d *Delegate) error {
+	if _, err := address.FromString(d.Owner); err != nil {
+		return errors.Wrap(err, "invalid owner address")
+	}
+
+	if _, err := address.FromString(d.Address); err != nil {
+		return errors.Wrap(err, "invalid operator address")
+	}
+
+	if _, err := address.FromString(d.RewardAddress); err != nil {
+		return errors.Wrap(err, "invalid reward address")
+	}
+
+	if d.Votes.Sign() <= 0 {
+		return ErrInvalidAmount
+	}
+	return nil
+}
+
+func (p *Protocol) addDelegate(d *Delegate) error {
+	if err := p.validateDelegate(d); err != nil {
+		return err
+	}
+
+	if !p.hasDelegate(d.CanName) {
```

Comment:
this function is not well defined. it should either "return error if canName is duplicate" or "replace the existing one"

Reply:
It is hard to review a private function because there is no context of how it will be used, and the functionality is conflict with the function name.

Reply:
this needs to be change to use Owner address